### PR TITLE
tests: Remove explicit uint64_t cast from Value

### PR DESCRIPTION
**Pull request by axic**

### CODE REVIEW

File: test/unittests/parser_test.cpp
Comment (axic) [inline test/unittests/parser_test.cpp:672]: This fails tests.

File: test/unittests/stack_test.cpp
Comment (axic) [inline test/unittests/stack_test.cpp:284]: Tried `ElementsAre(fizzy::Value{1}, ...)` but that complained there's no `operator=(const Value, const Value)` and it had to be fully qualified with `fizzy::` because otherwise conflicted with `Value` in gmock/gtest -- all in all, this is simpler.

File: test/unittests/parser_test.cpp
Comment (gumb0) [reply test/unittests/parser_test.cpp:672]: This is an index of `GlobalGet`, it should be `.i64`

File: test/unittests/api_test.cpp
Comment (gumb0) [inline test/unittests/api_test.cpp:318]: You change the test here to check only 32-bit, I thinks it's better to check `.i64`

Comment (axic) [reply test/unittests/api_test.cpp:318]: Changed it to i32 because that's what the sources use above.

File: test/unittests/parser_test.cpp
Comment (axic) [reply test/unittests/parser_test.cpp:672]: The other cases above use `f32` and `f64`.

Comment (axic) [reply test/unittests/parser_test.cpp:672]: Ah never mind, didn't see the `GlobalGet` vs `Constant`, will fix.

Comment (axic) [reply test/unittests/parser_test.cpp:672]: Btw, the test below uses `value.global_index` and not `value.constant` for that case, so I think this test here is a bit wrong.

Comment (gumb0) [reply test/unittests/parser_test.cpp:672]: Oh right, should be `value.global_index`

Comment (axic) [reply test/unittests/parser_test.cpp:672]: Fixed.

**Comment by axic** [comment]
Should squash the separate commits of `test` before merging.

File: test/unittests/execute_call_test.cpp
Comment (gumb0) [inline test/unittests/execute_call_test.cpp:291]: A bit hard to read
```suggestion
        const auto x = args[0].as<uint32_t>();
        return Value{uint64_t{x} * x};
```

File: test/unittests/api_test.cpp
Comment (gumb0) [reply test/unittests/api_test.cpp:318]: In case `find_exported_global` would return some incorrect `i64` global with the equal lower word, we would miss it.

File: test/unittests/instantiate_test.cpp
Comment (gumb0) [inline test/unittests/instantiate_test.cpp:22]: Here, too: if table function would return incorrect `i64` value, we could miss it, if we check only lower 32 bit.

**Comment by axic** [comment]
@gumb0 this is still waiting for a thorough review from you, I'd like to merge this.

File: test/unittests/api_test.cpp
Comment (chfast) [reply test/unittests/api_test.cpp:318]: I'm for staying with `i64` just because it tests more bits and goes along with the current implementation.

Comment (axic) [reply test/unittests/api_test.cpp:318]: With this reasoning, are we also using `.i64` for inputs even if the function type is defined as i32?

Comment (chfast) [reply test/unittests/api_test.cpp:318]: > With this reasoning, are we also using `.i64` for inputs even if the function type is defined as i32?

Not for computation, but you can add asserts there.

Comment (axic) [reply test/unittests/api_test.cpp:318]: That is a tad inconsistent. Perhaps the best really is a helper which asserts there are no dirty higher bits.

Comment (chfast) [reply test/unittests/api_test.cpp:318]: Testing more than is defined by wasm type system is fine to me.

File: test/utils/asserts.hpp
Comment (chfast) [inline test/utils/asserts.hpp:69]:
```suggestion
    EXPECT_EQ(value.i64 & 0xffffffff00000000, 0);
```
Yes, this looks good now. The suggestion should work too I think.

Alternatively, we may consider to add the assert to `Value::as<uint32_t>()`.

Comment (axic) [reply test/utils/asserts.hpp:69]: I had it without the postfix, but wanted to be sure. Perhaps not needed in C++.

Comment (axic) [reply test/utils/asserts.hpp:69]: > Alternatively, we may consider to add the assert to Value::as<uint32_t>().

Maybe we should do that :wink:

Comment (axic) [reply test/utils/asserts.hpp:69]: Added it, but we already have a trap with the unit tests:
```
[ RUN      ] end_to_end.milestone2
Assertion failed: ((i64 & 0xffffffff00000000) == 0), function as, file /Users/alex/Projects/fizzy/lib/fizzy/value.hpp, line 59.
Abort trap: 6
```

File: test/unittests/execute_call_test.cpp
Comment (axic) [inline test/unittests/execute_call_test.cpp:296]: These two (sqr, isqrt) expect i64 as output, should we explicitly cast them?

Comment (chfast) [reply test/unittests/execute_call_test.cpp:296]: No. This is fine.

File: test/unittests/parser_test.cpp
Comment (axic) [inline test/unittests/parser_test.cpp:672]: Hm, this was a typo iirc, we did not fix it in another PR? Will make sure this has its own commit.

File: test/unittests/stack_test.cpp
Comment (axic) [reply test/unittests/stack_test.cpp:284]: Could also add check for size of `result` to equal 3 here.

File: test/utils/asserts.hpp
Comment (chfast) [inline test/utils/asserts.hpp:67]: This is not used anywhere.

File: test/unittests/parser_test.cpp
Comment (chfast) [inline test/unittests/parser_test.cpp:631]:
```suggestion
    EXPECT_EQ(as_uint32(module.globalsec[1].expression.value.constant), uint32_t(-1));
```
This should stay as otherwise implicit conversion is needed.

File: test/unittests/parser_test.cpp
Comment (axic) [reply test/unittests/parser_test.cpp:672]: We actually have its own commit already.

File: test/unittests/parser_test.cpp
Comment (axic) [reply test/unittests/parser_test.cpp:631]: Couldn't we just use `as_int32(..), -1` ?

Comment (chfast) [reply test/unittests/parser_test.cpp]: In theory you can, but I don't see strong point for it. I prefer to refer to `i32` as `uint32_t` unless some heavy signed arithmetic is involved. Here you can just replace it with `0xffffffff`.

Comment (axic) [reply test/unittests/parser_test.cpp]: Alright, removed it.

File: test/utils/asserts.hpp
Comment (axic) [reply test/utils/asserts.hpp]: Created an issue to track this.

File: test/unittests/api_test.cpp
Comment (axic) [reply test/unittests/api_test.cpp]: Solved this by adding the helper.

File: test/unittests/stack_test.cpp
Comment (axic) [reply test/unittests/stack_test.cpp:284]: Done.

**Review by chfast — APPROVED** [review]

**Pull request merged**
